"corfu_stream write_stream" throws exception when connecting to failed server

Found on master at commit ee47f10ea05dcadb68ef9d7dd571510034b1ea6a

Create a cluster of two nodes and bootstrap with two sequencers, one layout, and one log server.  Kill one of the layout servers (the one that isn't providing sequencer or log service according to the bootstrapped config), then try to append to a stream using the remaining server.

    ./bin/corfu_server -l /tmp/buggy.8000 -a localhost -t -1 -d INFO 8000 &
    spid0=$!
    ./bin/corfu_server -l /tmp/buggy.8001 -a localhost -t -1 -d INFO 8001 &
    spid1=$!
    sleep 3
    rm -f /tmp/buggy.config
    cat<<! > /tmp/buggy.config
    {
        "layoutServers": [ "localhost:8000", "localhost:8001" ],
        "sequencers": [ "localhost:8000" ],
        "segments": [
        {
            "replicationMode": "CHAIN_REPLICATION",
            "start": 0,
             "end": -1,
             "stripes": [
             {
                 "logServers": [ "localhost:8000" ]
             }
             ]}],
        "epoch": 0
    }
    !
    ./bin/corfu_layout corfu_layout bootstrap -l /tmp/buggy.config localhost:8000
    ./bin/corfu_layout corfu_layout bootstrap -l /tmp/buggy.config localhost:8001
    
    echo
    echo "To kill the server on port 8000, use: pkill -P $spid0"
    
    echo "Now kill the server on port 8001"
    pkill -P $spid1
    
    echo "Try to write to a stream using server on port 8000"
    echo Hello, world | ./bin/corfu_stream corfu_stream write -c localhost:8000 -i test
    
    ## If the 'corfu_stream' thingie didn't hang forever

The cmdlet will retry every 5 seconds in an infinite loop, logging these messages:

    16:06:17.877 WARN  [ForkJoinPool.commonPool-worker-1] o.c.r.CorfuRuntime - Tried to get layout from localhost:8000 but failed with exception:
    org.corfudb.runtime.exceptions.NetworkException: ConnectException connecting to endpoint [endpoint=localhost:8001]
            at org.corfudb.runtime.clients.NettyClientRouter.start(NettyClientRouter.java:191) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at org.corfudb.runtime.CorfuRuntime.lambda$new$0(CorfuRuntime.java:108) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[na:1.8.0_91]
            at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1540) ~[na:1.8.0_91]
            at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[na:1.8.0_91]
            at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[na:1.8.0_91]
            at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[na:1.8.0_91]
            at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[na:1.8.0_91]
            at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:1.8.0_91]
            at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) ~[na:1.8.0_91]
            at org.corfudb.runtime.CorfuRuntime.lambda$fetchLayout$2(CorfuRuntime.java:219) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590) ~[na:1.8.0_91]
            at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1582) ~[na:1.8.0_91]
            at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[na:1.8.0_91]
            at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[na:1.8.0_91]
            at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) ~[na:1.8.0_91]
            at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[na:1.8.0_91]
    Caused by: java.net.ConnectException: Connection refused: localhost/127.0.0.1:8001
            at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) ~[na:1.8.0_91]
            at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717) ~[na:1.8.0_91]
            at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:224) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:289) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112) ~[corfu-0.1-SNAPSHOT-shaded.jar:na]
            at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_91]
    16:06:17.878 WARN  [ForkJoinPool.commonPool-worker-1] o.c.r.CorfuRuntime - Couldn't connect to any layout servers, retrying in 5s.

--- a/src/main/resources/logback.xml
+++ b/src/main/resources/logback.xml
@@ -2,7 +2,7 @@
 <configuration>
     <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
         <encoder>
-            <pattern>%d{HH:mm:ss.SSS} %highlight(%-5level) [%thread] %cyan(%logger{15}) - %msg%n</pattern>
+            <pattern>%d{HH:mm:ss.SSS} %highlight(%-5level) [%thread] %cyan(%logger{15}) - %msg%n %ex{short}</pattern>
         </encoder>
     </appender>
 

--- a/src/test/resources/logback-test.xml
+++ b/src/test/resources/logback-test.xml
@@ -2,7 +2,7 @@
 <configuration>
     <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
         <encoder>
-            <pattern>%d{HH:mm:ss.SSS} %highlight(%-5level) [%thread] %cyan(%logger{15}) - %msg%n</pattern>
+            <pattern>%d{HH:mm:ss.SSS} %highlight(%-5level) [%thread] %cyan(%logger{15}) - %msg%n %ex{short}</pattern>
         </encoder>
     </appender>
 

